From 7cce1e27a12fbbe978f4e351345e9161cb7d718a Mon Sep 17 00:00:00 2001
From: Vishal Sagar <vishal.sagar@xilinx.com>
Date: Thu, 3 Sep 2020 12:05:46 -0700
Subject: [PATCH 1552/1852] v4l: xilinx: sdirxss: Decode colorimetry from ST352

commit 3f204c1a4e3f534e02e5a733159f8b2b943db67c from
https://github.com/Xilinx/linux-xlnx.git

Update the colorimetry for the v4l2_mbus_format based on the ST352
payload. For SD mode, the colorimetry is SMPTE 170M. Otherwise it is BT
709. For HDR it is BT.2020. Otherwise it is set as default.

Signed-off-by: Vishal Sagar <vishal.sagar@xilinx.com>
Reviewed-by: Hyun Kwon <hyun.kwon@xilinx.com>
State: pending
Signed-off-by: Yaliang Wang <Yaliang.Wang@windriver.com>
---
 .../media/platform/xilinx/xilinx-sdirxss.c    | 28 +++++++++++++++++++
 1 file changed, 28 insertions(+)

diff --git a/drivers/media/platform/xilinx/xilinx-sdirxss.c b/drivers/media/platform/xilinx/xilinx-sdirxss.c
index 4c4f4a6fe3db..a47fc2d15922 100644
--- a/drivers/media/platform/xilinx/xilinx-sdirxss.c
+++ b/drivers/media/platform/xilinx/xilinx-sdirxss.c
@@ -254,6 +254,13 @@
 #define XST352_BYTE2_EOTF_HLG			0x1
 #define XST352_BYTE2_EOTF_SMPTE2084		0x2
 
+#define XST352_BYTE2_COLORIMETRY_MASK		GENMASK(21, 20)
+#define XST352_BYTE2_COLORIMETRY_OFFSET		20
+#define XST352_BYTE2_COLORIMETRY_BT709		0
+#define XST352_BYTE2_COLORIMETRY_VANC		1
+#define XST352_BYTE2_COLORIMETRY_UHDTV		2
+#define XST352_BYTE2_COLORIMETRY_UNKNOWN	3
+
 #define XST352_BYTE3_ACT_LUMA_COUNT_MASK	BIT(22)
 #define XST352_BYTE3_ACT_LUMA_COUNT_OFFSET	22
 
@@ -1347,11 +1354,15 @@ static int xsdirx_get_stream_properties(struct xsdirxss_state *state)
 	memset(&state->static_hdr, 0, sizeof(state->static_hdr));
 
 	state->static_hdr.eotf = V4L2_EOTF_TRADITIONAL_GAMMA_SDR;
+	format->colorspace = V4L2_COLORSPACE_SMPTE170M;
 
 	if (mode != XSDIRX_MODE_SD_MASK) {
 		u8 eotf = (payload & XST352_BYTE2_EOTF_MASK) >>
 			XST352_BYTE2_EOTF_OFFSET;
 
+		u8 colorimetry = (payload & XST352_BYTE2_COLORIMETRY_MASK) >>
+			XST352_BYTE2_COLORIMETRY_OFFSET;
+
 		/* Get the EOTF function */
 		switch (eotf) {
 		case XST352_BYTE2_EOTF_SDRTV:
@@ -1365,6 +1376,23 @@ static int xsdirx_get_stream_properties(struct xsdirxss_state *state)
 			state->static_hdr.eotf = V4L2_EOTF_BT_2100_HLG;
 			break;
 		}
+
+		/* Get the colorimetry data */
+		switch (colorimetry) {
+		case XST352_BYTE2_COLORIMETRY_BT709:
+			format->colorspace = V4L2_COLORSPACE_REC709;
+			break;
+		case XST352_BYTE2_COLORIMETRY_UHDTV:
+			format->colorspace = V4L2_COLORSPACE_BT2020;
+			break;
+		default:
+			/*
+			 * Modes which will have VANC and Unknown colorimetery
+			 * are currently not supported
+			 */
+			format->colorspace = V4L2_COLORSPACE_DEFAULT;
+			break;
+		}
 	}
 
 	dev_dbg(core->dev, "Stream width = %d height = %d Field = %d payload = 0x%08x ts = 0x%08x\n",
-- 
2.31.1

